src/util/result.h

```diff
@@ -215,29 +294,59 @@ class Result : public detail::SuccessHolder<SuccessType_, FailureType_, Messages
     template <bool Failure, typename Result, typename... Args>
     static void Construct(Result& result, util::Error error, Args&&... args)
     {
-        result.EnsureFailData().messages = std::move(error.message);
+        result.AddError(std::move(error.message));
         Construct</*Failure=*/true>(result, std::forward<Args>(args)...);
     }
 
+    //! Construct() overload peeling off a util::Warning constructor argument.
+    template <bool Failure, typename Result, typename... Args>
+    static void Construct(Result& result, util::Warning warning, Args&&... args)
+    {
+        result.AddWarning(std::move(warning.message));
+        Construct<Failure>(result, std::forward<Args>(args)...);
+    }
+
+    //! Construct() overload peeling off a util::MoveFrom constructor argument.
+    template <bool Failure, typename Result, typename R, typename... Args>
+    static void Construct(Result& result, MoveFrom<R> other, Args&&... args)
+    {
+        other.m_result >> result;
+        Construct<Failure>(result, std::forward<Args>(args)...);
+    }
+
     //! Move success, failure, and messages from source Result object to
-    //! destination object. The source and destination results are not required
+    //! destination object. Existing values are merged (using
+    //! ResultTraits::MergeInto specializations), so destination errors and
+    //! warning messages get appended to instead of overwritten.
+    //! The source and destination results are not required
     //! to have the same types, and assigning void source types to non-void
     //! destinations type is allowed, since no source information is lost. But
     //! assigning non-void source types to void destination types is not
     //! allowed, since this would discard source information.
     template <bool DstConstructed, typename DstResult, typename SrcResult>
     static void Move(DstResult& dst, SrcResult& src)
     {
-        // Move messages values first, then move success or failure value below.
-        if (src.GetMessages() && !src.GetMessages()->empty()) {
-            dst.EnsureMessages() = std::move(*src.GetMessages());
-        }
+        // Use operator>> to move messages values first, then move success or
+        // failure value below.
+        src >> dst;
         // If DstConstructed is true, it means dst has either a success value or
-        // a failure value set, which needs to be destroyed and replaced. If
+        // a failure value set, which needs to be merged or replaced. If
         // DstConstructed is false, then dst is being constructed now and has no
         // values set.
         if constexpr (DstConstructed) {
-            if (dst) {
+            if (dst && src) {
+                // dst and src both hold success values, so merge them and return
+                if constexpr (!std::is_same_v<typename SrcResult::SuccessType, void>) {
+                    ResultTraits<typename SrcResult::SuccessType>::MergeInto(*dst, *src);
```

Comment:
Reviewers might be wondering what the benefits of  the ResultTraits are. They are useful for potential future extensions of the move behaviour: https://github.com/bitcoin/bitcoin/pull/29700/commits/3951afc3b708326cea653951ef331d8f96a28682 .